<a href="https://colab.research.google.com/github/SwathiMenon/Machine_learning/blob/main/Personality_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Aim: The  aim is to classify how likely someone will make friends based on their social skills.

Dataset Name:**Personality classification Type Dataset**


Dataset link: https://www.kaggle.com/datasets/shariful07/personality-classification-type-dataset

In [35]:
#Importing The Required Libraries
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from sklearn.feature_selection import SelectKBest
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from xgboost import XGBClassifier
import warnings
warnings.filterwarnings("ignore")
from sklearn.metrics import classification_report,ConfusionMatrixDisplay

In [2]:
df=pd.read_csv("/content/drive/MyDrive/datasets/Personality classification Data .csv")
df

Worry about things.  Make friends easily.  Have a vivid imagination.  \
0                   -1.0                   0.0                        1.0   
1                    0.0                   1.0                        1.0   
2                   -1.0                   1.0                        1.0   
3                   -2.0                   1.0                        0.0   
4                   -1.0                   1.0                        1.0   
..                   ...                   ...                        ...   
117                  NaN                   1.0                        1.0   
118                 -1.0                   1.0                        0.0   
119                 -1.0                   1.0                        1.0   
120                 -2.0                   0.0                        1.0   
121                 -1.0                   1.0                        1.0   

     Trust others  Get angry easily.  Love large parties.  \
0               1               -1.0                    0   
1               2                0.0                    1   
2               1               -1.0                    1   
3               0               -2.0                    2   
4               1               -1.0                    1   
..            ...                ...                  ...   
117             1               -1.0                    1   
118             2               -1.0                    0   
119             0               -2.0                    1   
120             0               -2.0                    2   
121             1               -1.0                    1   

     Would never cheat to get ahead.  Like order.  Often feel unhappy.  \
0                               -1.0            1                  0.0   
1                               -2.0            2                 -1.0   
2                               -1.0            1                  0.0   
3                               -1.0            0                  0.0   
4                               -2.0            1                  1.0   
..                               ...          ...                  ...   
117                             -2.0            1                  1.0   
118                             -2.0            2                 -1.0   
119                             -1.0            0                  1.0   
120                             -2.0            0                  1.0   
121                             -1.0            1                  0.0   

     Take charge.  ...  Stick to the rules.  Try to lead others.  \
0               1  ...                  1.0                    1   
1               2  ...                  1.0                    1   
2               1  ...                  2.0                    1   
3               0  ...                  1.0                    1   
4               1  ...                  0.0                    1   
..            ...  ...                  ...                  ...   
117             1  ...                  1.0                    1   
118             2  ...                  0.0                    0   
119             0  ...                  2.0                    1   
120             0  ...                  2.0                    1   
121             1  ...                  1.0                    0   

     Keep my promises.  Feel others' emotions  Like to visit new places.  \
0                    2                    1.0                          2   
1                    1                    2.0                          2   
2                    0                    1.0                          2   
3                    1                    0.0                          1   
4                    2                    1.0                          2   
..                 ...                    ...                        ...   
117                  1                    2.0                          2   
118                  2           

# **Preprocessing**

In [3]:
df.isna().sum()

Worry about things.                 5
Make friends easily.                5
Have a vivid imagination.           2
Trust others                        0
Get angry easily.                   2
Love large parties.                 0
Would never cheat to get ahead.     5
Like order.                         0
Often feel unhappy.                 2
Take charge.                        0
Make people feel welcome.           1
Try to follow the rules.            0
Am always busy.                     3
Am easy to satisfy.                 0
Go straight for the goal.           0
Often overindulge.                  3
Panic easily.                       7
Avoid mistakes.                     0
Warm up quickly to others.          4
Get irritated easily.               6
Stick to the rules.                 6
Try to lead others.                 0
Keep my promises.                   0
Feel others' emotions               2
Like to visit new places.           0
Work hard.                         13
Become overw

In [4]:
lst=["Worry about things.","Make friends easily.","Have a vivid imagination.","Get angry easily.","Would never cheat to get ahead.","Often feel unhappy.","Make people feel welcome.","Am always busy.","Often overindulge.","Panic easily.","Warm up quickly to others.","Get irritated easily.","Stick to the rules.","Feel others' emotions","Work hard.","I have little to say.","I am quiet around strangers."]
for i in lst:
  df[i]=df[i].fillna(df[i].mode()[0])

In [5]:
df.isna().sum()

Worry about things.                0
Make friends easily.               0
Have a vivid imagination.          0
Trust others                       0
Get angry easily.                  0
Love large parties.                0
Would never cheat to get ahead.    0
Like order.                        0
Often feel unhappy.                0
Take charge.                       0
Make people feel welcome.          0
Try to follow the rules.           0
Am always busy.                    0
Am easy to satisfy.                0
Go straight for the goal.          0
Often overindulge.                 0
Panic easily.                      0
Avoid mistakes.                    0
Warm up quickly to others.         0
Get irritated easily.              0
Stick to the rules.                0
Try to lead others.                0
Keep my promises.                  0
Feel others' emotions              0
Like to visit new places.          0
Work hard.                         0
Become overwhelmed by events.      0
C

In [6]:
df.dtypes

Worry about things.                float64
Make friends easily.               float64
Have a vivid imagination.          float64
Trust others                         int64
Get angry easily.                  float64
Love large parties.                  int64
Would never cheat to get ahead.    float64
Like order.                          int64
Often feel unhappy.                float64
Take charge.                         int64
Make people feel welcome.          float64
Try to follow the rules.             int64
Am always busy.                    float64
Am easy to satisfy.                  int64
Go straight for the goal.            int64
Often overindulge.                 float64
Panic easily.                      float64
Avoid mistakes.                      int64
Warm up quickly to others.         float64
Get irritated easily.              float64
Stick to the rules.                float64
Try to lead others.                  int64
Keep my promises.                    int64
Feel others

In [7]:
x=df.drop(["Make friends easily."],axis=1)
x

Worry about things.  Have a vivid imagination.  Trust others  \
0                   -1.0                        1.0             1   
1                    0.0                        1.0             2   
2                   -1.0                        1.0             1   
3                   -2.0                        0.0             0   
4                   -1.0                        1.0             1   
..                   ...                        ...           ...   
117                 -1.0                        1.0             1   
118                 -1.0                        0.0             2   
119                 -1.0                        1.0             0   
120                 -2.0                        1.0             0   
121                 -1.0                        1.0             1   

     Get angry easily.  Love large parties.  Would never cheat to get ahead.  \
0                 -1.0                    0                             -1.0   
1                  0.0                    1                             -2.0   
2                 -1.0                    1                             -1.0   
3                 -2.0                    2                             -1.0   
4                 -1.0                    1                             -2.0   
..                 ...                  ...                              ...   
117               -1.0                    1                             -2.0   
118               -1.0                    0                             -2.0   
119               -2.0                    1                             -1.0   
120               -2.0                    2                             -2.0   
121               -1.0                    1                             -1.0   

     Like order.  Often feel unhappy.  Take charge.  \
0              1                  0.0             1   
1              2                 -1.0             2   
2              1                  0.0             1   
3              0                  0.0             0   
4              1                  1.0             1   
..           ...                  ...           ...   
117            1                  1.0             1   
118            2                 -1.0             2   
119            0                  1.0             0   
120            0                  1.0             0   
121            1                  0.0             1   

     Make people feel welcome.  ...  Stick to the rules.  Try to lead others.  \
0                          0.0  ...                  1.0                    1   
1                          1.0  ...                  1.0                    1   
2                          1.0  ...                  2.0                    1   
3                          1.0  ...                  1.0                    1   
4                          1.0  ...                  0.0                    1   
..                         ...  ...                  ...                  ...   
117                        1.0  ...                  1.0                    1   
118                        0.0  ...                  0.0                    0   
119                        1.0  ...                  2.0                    1   
120                        0.0  ...                  2.0                    1   
121                        1.0  ...                  1.0                    0   

     Keep my promises.  Feel others' emotions  Like to visit new places.  \
0                    2                    1.0                          2   
1                    1                    2.0                          2   
2                    0                    1.0                          2   
3                    1                    0.0                          1   
4                    2                    1.0                          2   
..                 ...                    ...                        ...   
117                  1                    2.0    

In [8]:
y=df.iloc[:,1]
y

0      0.0
1      1.0
2      1.0
3      1.0
4      1.0
      ... 
117    1.0
118    1.0
119    1.0
120    0.0
121    1.0
Name: Make friends easily., Length: 122, dtype: float64

In [9]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.5,random_state=1)

In [10]:
#scaling
sc=StandardScaler()
x_train=sc.fit_transform(x_train)
x_test=sc.transform(x_test)

In [11]:
y.value_counts()

1.0    87
0.0    35
Name: Make friends easily., dtype: int64

In [12]:
#random oversampling is done
from imblearn.over_sampling import RandomOverSampler
sm=RandomOverSampler()
xtrain,ytrain=sm.fit_resample(x_train,y_train)

# **Model Building**

In [14]:
#KNN
knn=KNeighborsClassifier()
knn.fit(x_train,y_train)
y_pred=knn.predict(x_test)
y_pred

array([1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 1., 1., 1., 1., 0., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1., 0.,
       1., 0., 1., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 0., 1., 1., 1., 0., 0., 1.])

In [16]:
#hyper tuning for knn
from sklearn.model_selection import GridSearchCV
params={'n_neighbors':[3,5,7,9],'weights':['uniform','distance'],
        'algorithm':['auto', 'ball_tree', 'kd_tree', 'brute']}
clf=GridSearchCV(knn,params,cv=5,scoring='accuracy')
clf.fit(x_train,y_train)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'n_neighbors': [3, 5, 7, 9],
                         'weights': ['uniform', 'distance']},
             scoring='accuracy')

In [17]:
print(clf.best_params_)

{'algorithm': 'auto', 'n_neighbors': 5, 'weights': 'distance'}


In [19]:
knn1=KNeighborsClassifier(algorithm='auto',n_neighbors=9,weights='distance')
knn1.fit(x_train,y_train)
y_pred1=knn1.predict(x_test)
print(classification_report(y_test,y_pred1))

              precision    recall  f1-score   support

         0.0       0.47      0.57      0.52        14
         1.0       0.86      0.81      0.84        47

    accuracy                           0.75        61
   macro avg       0.67      0.69      0.68        61
weighted avg       0.77      0.75      0.76        61



In [22]:
#svc


param_grid = {'C': [0.1, 1, 10], 'kernel': ['linear', 'rbf'], 'gamma': [0.1, 1, 10]}

# Create an SVC classifier
svc = SVC()

# Create a GridSearchCV object and fit it to the training data
grid_search = GridSearchCV(estimator=svc, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(x_train, y_train)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train a new SVC model with the best parameters
best_svc = SVC(**best_params)
best_svc.fit(x_train, y_train)

# Evaluate the model on the testing set
accuracy = best_svc.score(x_test, y_test)

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Test Accuracy:", accuracy)



Best Parameters: {'C': 0.1, 'gamma': 0.1, 'kernel': 'linear'}
Best Score: 0.8679487179487179
Test Accuracy: 0.7049180327868853


In [24]:
sv=SVC(C=10, gamma =1, kernel= 'rbf')
sv.fit(x_train,y_train)
y_pred2=sv.predict(x_test)
y_pred2

array([1., 1., 1., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 0., 1., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 0., 1., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 0., 1.])

In [25]:
print(classification_report(y_test,y_pred2))

              precision    recall  f1-score   support

         0.0       0.73      0.57      0.64        14
         1.0       0.88      0.94      0.91        47

    accuracy                           0.85        61
   macro avg       0.80      0.75      0.77        61
weighted avg       0.84      0.85      0.85        61



In [27]:
#Gaussian NB
nb=GaussianNB()
nb.fit(x_train,y_train)
y_pred2=nb.predict(x_test)
y_pred2

array([0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 1., 1., 1., 0., 0., 0., 1.,
       1., 0., 0., 1., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 0.,
       1., 1., 0., 1., 1., 1., 0., 0., 0., 1., 1., 1., 1., 1., 1., 0., 0.,
       0., 1., 1., 0., 0., 0., 1., 1., 0., 1.])

In [28]:
print(classification_report(y_test,y_pred2))

              precision    recall  f1-score   support

         0.0       0.35      0.79      0.49        14
         1.0       0.90      0.57      0.70        47

    accuracy                           0.62        61
   macro avg       0.63      0.68      0.60        61
weighted avg       0.77      0.62      0.65        61



In [29]:
#DecisionTreeClassifier
#hyper tuning
# Define the parameter grid for GridSearchCV
param_grid = {'criterion': ['gini', 'entropy'], 'max_depth': [None, 5, 10, 15]}

# Create a DecisionTreeClassifier
dtc = DecisionTreeClassifier()

# Create a GridSearchCV object and fit it to the training data
grid_search = GridSearchCV(estimator=dtc, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(x_train, y_train)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train a new DecisionTreeClassifier with the best parameters
best_dtc = DecisionTreeClassifier(**best_params)
best_dtc.fit(x_train, y_train)

# Evaluate the model on the testing set
accuracy = best_dtc.score(x_test, y_test)

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Test Accuracy:", accuracy)

Best Parameters: {'criterion': 'gini', 'max_depth': 5}
Best Score: 0.7897435897435898
Test Accuracy: 0.7704918032786885


In [37]:
dt=DecisionTreeClassifier(criterion='gini',random_state=2,max_depth=5)
dt.fit(x_train,y_train)
y_pred3=dt.predict(x_test)
y_pred3

array([1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 1., 0.,
       1., 1., 0., 1., 0., 1., 1., 0., 1., 0., 1., 0., 1., 0., 1., 1., 1.,
       1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 1., 0., 1.,
       0., 1., 1., 0., 0., 1., 1., 1., 0., 1.])

In [32]:
print(classification_report(y_test,y_pred3))

              precision    recall  f1-score   support

         0.0       0.50      0.93      0.65        14
         1.0       0.97      0.72      0.83        47

    accuracy                           0.77        61
   macro avg       0.74      0.83      0.74        61
weighted avg       0.86      0.77      0.79        61



In [36]:
#Random Forest
#hyper tuning
param_grid = {'n_estimators': [100, 200, 300], 'criterion': ['gini', 'entropy'],'max_depth': [None, 5, 10],'min_samples_split': [2, 4],'min_samples_leaf': [1, 2]}

# Create a RandomForestClassifier
rfc = RandomForestClassifier()

# Create a GridSearchCV object and fit it to the training data
grid_search = GridSearchCV(estimator=rfc, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(x_train, y_train)
# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train a new RandomForestClassifier with the best parameters
best_rfc = RandomForestClassifier(**best_params)
best_rfc.fit(x_train, y_train)

# Evaluate the model on the testing set
accuracy = best_rfc.score(x_test, y_test)

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Test Accuracy:", accuracy)

Best Parameters: {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best Score: 0.8525641025641025
Test Accuracy: 0.8524590163934426


In [42]:
dt=DecisionTreeClassifier(criterion='entropy',random_state=3,max_depth=5)
dt.fit(x_train,y_train)
y_pred3=dt.predict(x_test)
y_pred3

array([1., 0., 1., 1., 0., 0., 0., 0., 0., 1., 1., 1., 0., 0., 1., 1., 0.,
       1., 1., 0., 1., 0., 1., 1., 0., 1., 1., 1., 0., 0., 1., 1., 1., 1.,
       1., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 1., 0., 1.,
       1., 1., 1., 1., 0., 1., 1., 1., 0., 1.])

In [43]:
print(classification_report(y_test,y_pred3))

              precision    recall  f1-score   support

         0.0       0.50      0.93      0.65        14
         1.0       0.97      0.72      0.83        47

    accuracy                           0.77        61
   macro avg       0.74      0.83      0.74        61
weighted avg       0.86      0.77      0.79        61



In [45]:
#RandomForestClassifier
#hyper tuning
param_grid = {'n_estimators': [100, 200, 300], 'criterion': ['gini', 'entropy'],'max_depth': [None, 5, 10],'min_samples_split': [2, 4],'min_samples_leaf': [1, 2]}

# Create a RandomForestClassifier
rfc = RandomForestClassifier()

# Create a GridSearchCV object and fit it to the training data
grid_search = GridSearchCV(estimator=rfc, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(x_train, y_train)
# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train a new RandomForestClassifier with the best parameters
best_rfc = RandomForestClassifier(**best_params)
best_rfc.fit(x_train, y_train)

# Evaluate the model on the testing set
accuracy = best_rfc.score(x_test, y_test)

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Test Accuracy:", accuracy)

Best Parameters: {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best Score: 0.8525641025641025
Test Accuracy: 0.8852459016393442


In [46]:
rf=RandomForestClassifier(criterion= 'entropy', max_depth= 5, min_samples_leaf= 1, min_samples_split= 2,n_estimators= 200)
rf.fit(x_train,y_train)
y_pred4=rf.predict(x_test)
y_pred4

array([1., 1., 1., 1., 0., 0., 1., 0., 0., 0., 1., 1., 1., 0., 1., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 0., 0., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0., 1.,
       1., 1., 1., 1., 0., 1., 1., 1., 0., 1.])

In [47]:
print(classification_report(y_test,y_pred4))

              precision    recall  f1-score   support

         0.0       0.71      0.86      0.77        14
         1.0       0.95      0.89      0.92        47

    accuracy                           0.89        61
   macro avg       0.83      0.88      0.85        61
weighted avg       0.90      0.89      0.89        61



In [49]:
#xgboost
# Define the parameter grid for GridSearchCV
param_grid = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.1, 0.5, 1.0],
    'max_depth': [3, 5, 7],
    'subsample': [0.8, 1.0],
    'colsample_bytree': [0.8, 1.0]
}

# Create an XGBClassifier
xgb = XGBClassifier()

# Create a GridSearchCV object and fit it to the training data
grid_search = GridSearchCV(estimator=xgb, param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(x_train, y_train)

# Get the best parameters and best score
best_params = grid_search.best_params_
best_score = grid_search.best_score_

# Train a new XGBClassifier with the best parameters
best_xgb = XGBClassifier(**best_params)
best_xgb.fit(x_train, y_train)

# Evaluate the model on the testing set
accuracy = best_xgb.score(x_test, y_test)

print("Best Parameters:", best_params)
print("Best Score:", best_score)
print("Test Accuracy:", accuracy)

Best Parameters: {'colsample_bytree': 1.0, 'learning_rate': 1.0, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.8}
Best Score: 0.8692307692307691
Test Accuracy: 0.7704918032786885


In [50]:
xgb=XGBClassifier()
xgb.fit(x_train,y_train)
y_pred7=xgb.predict(x_test)
y_pred7

array([1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1])

In [51]:
print(classification_report(y_test,y_pred7))

              precision    recall  f1-score   support

         0.0       0.50      0.86      0.63        14
         1.0       0.95      0.74      0.83        47

    accuracy                           0.77        61
   macro avg       0.72      0.80      0.73        61
weighted avg       0.84      0.77      0.79        61



higest accuracy is in random forest classifier with 89%